src/nbl/asset/utils/IShaderCompiler.cpp

```diff
@@ -289,3 +219,116 @@ auto IShaderCompiler::CIncludeFinder::tryIncludeGenerators(const std::string& in
+    retVal.insert(retVal.end(), reinterpret_cast<uint8_t*>(&dumpedContainerJsonLength), reinterpret_cast<uint8_t*>(&dumpedContainerJsonLength) + CONTAINER_JSON_SIZE_BYTES);
+    retVal.insert(retVal.end(), std::make_move_iterator(dumpedContainerJson.begin()), std::make_move_iterator(dumpedContainerJson.end()));
+    retVal.insert(retVal.end(), std::make_move_iterator(shadersBuffer.begin()), std::make_move_iterator(shadersBuffer.end()));
+    return core::make_smart_refctd_ptr<CVectorCPUBuffer<uint8_t, nbl::core::aligned_allocator<uint8_t>>>(std::move(retVal));
+}
+
+core::smart_refctd_ptr<IShaderCompiler::CCache> IShaderCompiler::CCache::deserialize(const std::span<const uint8_t> serializedCache)
+{
+    auto retVal = core::make_smart_refctd_ptr<CCache>();
+
+    // First get the size of the json in the buffer, stored in the first 8 bytes
+    const uint64_t* cacheStart = reinterpret_cast<const uint64_t*>(serializedCache.data());
+    uint64_t containerJsonSize = cacheStart[0];
+    // Next up get the json that stores the container data
+    std::span<const char> cacheAsChar = { reinterpret_cast<const char*>(serializedCache.data()), serializedCache.size() };
+    std::string_view containerJsonString(cacheAsChar.begin() + CONTAINER_JSON_SIZE_BYTES, cacheAsChar.begin() + CONTAINER_JSON_SIZE_BYTES + containerJsonSize);
+    json containerJson = json::parse(containerJsonString);
+    
+    // Now retrieve two vectors, one with the entries and one with the extra data to recreate the CPUShaders
+    std::vector<SEntry> entries;
+    std::vector<CPUShaderCreationParams> shaderCreationParams;
+    containerJson.at("entries").get_to(entries);
+    containerJson.at("shaderCreationParams").get_to(shaderCreationParams);
+
+    // We must now recreate the shaders, add them to each entry, then move the entry into the multiset
+    for (auto i = 0u; i < entries.size(); i++) {
+        // Create buffer to hold the code
+        auto code = core::make_smart_refctd_ptr<ICPUBuffer>(shaderCreationParams[i].codeByteSize);
+        // Copy the shader bytecode into the buffer
+        memcpy(code->getPointer(), serializedCache.data() + CONTAINER_JSON_SIZE_BYTES + containerJsonSize + shaderCreationParams[i].offset, shaderCreationParams[i].codeByteSize);
+        // Create the ICPUShader
+        auto value = core::make_smart_refctd_ptr<ICPUShader>(std::move(code), shaderCreationParams[i].stage, shaderCreationParams[i].contentType, std::move(shaderCreationParams[i].filepathHint));
+       
+        entries[i].value = std::move(value);
+
+        retVal->insert(std::move(entries[i]));
+    }
+
+    return retVal;
+}
+
+core::smart_refctd_ptr<IShaderCompiler::CCache> IShaderCompiler::CCache::deserialize(core::smart_refctd_ptr<const system::IFile> serializedCache) {
+    const void* fileBufferPointer = serializedCache->getMappedPointer();
+    assert(fileBufferPointer); // Check the file is readable
```

Comment:
just return `nullptr`, asserts make applications crash, they're for abnormal errors/assumptions

Also check the mapping flags for reading